File: samples/grpc/iot-service-scala/src/main/scala/iot/temperature/SensorTwinServiceImpl.scala
package iot.temperature

import java.util.concurrent.TimeoutException

import scala.concurrent.Future

import akka.actor.typed.ActorSystem
import akka.cluster.sharding.typed.scaladsl.ClusterSharding
import akka.grpc.GrpcServiceException
import akka.util.Timeout
import io.grpc.Status
import iot.temperature.proto.SensorTwinService

class SensorTwinServiceImpl(system: ActorSystem[_]) extends SensorTwinService {

  import system.executionContext

  implicit private val timeout: Timeout =
    Timeout.create(
      system.settings.config.getDuration("iot-service.ask-timeout"))

  private val sharding = ClusterSharding(system)

  override def getTemperature(
      in: proto.GetTemperatureRequest): Future[proto.CurrentTemperature] = {
    val entityRef = sharding.entityRefFor(SensorTwin.EntityKey, in.sensorEntityId)
    val reply: Future[Int] =
      entityRef.askWithStatus(SensorTwin.GetTemperature(_))
    val response =
      reply.map(temperature =>
        proto.CurrentTemperature(in.sensorEntityId, temperature))
    convertError(response)
  }

  private def convertError[T](response: Future[T]): Future[T] = {
    response.recoverWith {
      case _: TimeoutException =>
        Future.failed(
          new GrpcServiceException(
            Status.UNAVAILABLE.withDescription("Operation timed out")))
      case exc =>
        Future.failed(
          new GrpcServiceException(
            Status.INVALID_ARGUMENT.withDescription(exc.getMessage)))
    }
  }
}

File: samples/grpc/iot-service-java/src/main/java/iot/temperature/SensorTwinServiceImpl.java
package iot.temperature;

import akka.actor.typed.ActorSystem;
import akka.cluster.sharding.typed.javadsl.ClusterSharding;
import akka.cluster.sharding.typed.javadsl.EntityRef;
import akka.grpc.GrpcServiceException;
import io.grpc.Status;
import iot.temperature.proto.CurrentTemperature;
import iot.temperature.proto.GetTemperatureRequest;
import iot.temperature.proto.SensorTwinService;
import java.time.Duration;
import java.util.concurrent.CompletionStage;

public class SensorTwinServiceImpl implements SensorTwinService {
  private final ActorSystem<?> system;
  private final Duration timeout;
  private final ClusterSharding sharding;

  public SensorTwinServiceImpl(ActorSystem<?> system) {
    this.system = system;
    this.timeout = system.settings().config().getDuration("iot-service.ask-timeout");
    this.sharding = ClusterSharding.get(system);
  }

  @Override
  public CompletionStage<CurrentTemperature> getTemperature(GetTemperatureRequest in) {
    EntityRef<SensorTwin.Command> entityRef =
        sharding.entityRefFor(SensorTwin.ENTITY_KEY, in.getSensorEntityId());
    CompletionStage<Integer> response =
        entityRef.askWithStatus(SensorTwin.GetTemperature::new, timeout);
    return response.handle(
        (value, error) -> {
          if (error != null) {
            throw new GrpcServiceException(Status.INTERNAL.withDescription(error.getMessage()));
          } else {
            return CurrentTemperature.newBuilder().setTemperature(value).build();
          }
        });
  }
}
